<a href="https://colab.research.google.com/github/Mandar-alt/ArimaModel/blob/main/Volatility_Nifty500.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
nifty500_stocks = pd.read_csv('Nifty_500.csv')

In [ ]:
print(nifty500_stocks)

            Date          Open          High           Low         Close  \
0     2018-08-16   9696.900391   9742.500000   9666.500000   9685.650391   
1     2018-08-17   9730.849609   9776.299805   9728.200195   9770.200195   
2     2018-08-20   9800.799805   9846.549805   9798.900391   9839.750000   
3     2018-08-21   9862.900391   9868.500000   9831.049805   9862.500000   
4     2018-08-23   9900.049805   9900.049805   9839.000000   9872.549805   
...          ...           ...           ...           ...           ...   
1473  2024-08-09  22948.150391  22982.449219  22869.400391  22923.500000   
1474  2024-08-12  22862.900391  22993.250000  22753.750000  22915.849609   
1475  2024-08-13  22952.699219  22962.750000  22684.449219  22705.800781   
1476  2024-08-14  22765.650391  22778.949219  22599.949219  22672.650391   
1477  2024-08-16  22846.099609  22900.250000  22750.300781  22806.000000   

         Adj Close      Volume  
0      9685.637695  13633700.0  
1      9770.187500  1

In [ ]:
print(nifty500_stocks.head())

         Date         Open         High          Low        Close  \
0  2018-08-16  9696.900391  9742.500000  9666.500000  9685.650391   
1  2018-08-17  9730.849609  9776.299805  9728.200195  9770.200195   
2  2018-08-20  9800.799805  9846.549805  9798.900391  9839.750000   
3  2018-08-21  9862.900391  9868.500000  9831.049805  9862.500000   
4  2018-08-23  9900.049805  9900.049805  9839.000000  9872.549805   

     Adj Close      Volume  
0  9685.637695  13633700.0  
1  9770.187500  12926000.0  
2  9839.737305  12305300.0  
3  9862.487305  12707800.0  
4  9872.537109  10832500.0  


In [ ]:
print(nifty500_stocks[['Date', 'Adj Close']])


            Date     Adj Close
0     2018-08-16   9685.637695
1     2018-08-17   9770.187500
2     2018-08-20   9839.737305
3     2018-08-21   9862.487305
4     2018-08-23   9872.537109
...          ...           ...
1473  2024-08-09  22923.500000
1474  2024-08-12  22915.849609
1475  2024-08-13  22705.800781
1476  2024-08-14  22672.650391
1477  2024-08-16  22806.000000

[1478 rows x 2 columns]


In [ ]:
# Calculated simple return of the adjusted close column
nifty500_stocks['simple_return'] = nifty500_stocks['Adj Close'].pct_change()
print(nifty500_stocks[['Date', 'Adj Close', 'simple_return']])


            Date     Adj Close  simple_return
0     2018-08-16   9685.637695            NaN
1     2018-08-17   9770.187500       0.008729
2     2018-08-20   9839.737305       0.007119
3     2018-08-21   9862.487305       0.002312
4     2018-08-23   9872.537109       0.001019
...          ...           ...            ...
1473  2024-08-09  22923.500000       0.009592
1474  2024-08-12  22915.849609      -0.000334
1475  2024-08-13  22705.800781      -0.009166
1476  2024-08-14  22672.650391      -0.001460
1477  2024-08-16  22806.000000       0.005882

[1478 rows x 3 columns]


<ipython-input-18-da87da91fb21>:2: FutureWarning: The default fill_method='pad' in Series.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  nifty500_stocks['simple_return'] = nifty500_stocks['Adj Close'].pct_change()


Calculating Stock Volatility

In [ ]:

# Calculating Stock Volatility
nifty500_stocks['daily_std_dev'] = nifty500_stocks['simple_return'].std()
print(nifty500_stocks[['Date', 'Adj Close', 'daily_std_dev']])


            Date     Adj Close  daily_std_dev
0     2018-08-16   9685.637695       0.011398
1     2018-08-17   9770.187500       0.011398
2     2018-08-20   9839.737305       0.011398
3     2018-08-21   9862.487305       0.011398
4     2018-08-23   9872.537109       0.011398
...          ...           ...            ...
1473  2024-08-09  22923.500000       0.011398
1474  2024-08-12  22915.849609       0.011398
1475  2024-08-13  22705.800781       0.011398
1476  2024-08-14  22672.650391       0.011398
1477  2024-08-16  22806.000000       0.011398

[1478 rows x 3 columns]


In [ ]:
# Calculate annualized standard deviation
nifty500_stocks['annualized_std_dev'] = nifty500_stocks['daily_std_dev'] * np.sqrt(252)
print(nifty500_stocks[['Date', 'Adj Close', 'annualized_std_dev']])

            Date     Adj Close  annualized_std_dev
0     2018-08-16   9685.637695            0.180941
1     2018-08-17   9770.187500            0.180941
2     2018-08-20   9839.737305            0.180941
3     2018-08-21   9862.487305            0.180941
4     2018-08-23   9872.537109            0.180941
...          ...           ...                 ...
1473  2024-08-09  22923.500000            0.180941
1474  2024-08-12  22915.849609            0.180941
1475  2024-08-13  22705.800781            0.180941
1476  2024-08-14  22672.650391            0.180941
1477  2024-08-16  22806.000000            0.180941

[1478 rows x 3 columns]


In [ ]:
print("Annunal Stock volatilty for 0 index : {0}".format(nifty500_stocks['annualized_std_dev'][0]))


In [ ]:

# Set threshold for high and low volatility
high_volatility_threshold = 0.3
low_volatility_threshold = 0.1

# Classify stocks based on volatility
high_volatility_stocks = nifty500_stocks[nifty500_stocks['annualized_std_dev'] > high_volatility_threshold]
low_volatility_stocks = nifty500_stocks[nifty500_stocks['annualized_std_dev'] < low_volatility_threshold]

print("High Volatility Stocks: ")
print(high_volatility_stocks)

print("\nLow Volatility Stocks:")
print(low_volatility_stocks)


In [ ]:
import yfinance as yf
import numpy as np

In [ ]:
nifty_500 = ['ZOMATO.NS', 'AWL.NS', 'J&KBANK.NS', 'PAYTM.NS', 'PCBL.NS',
                     'JUBLINGREA.NS', 'NAUKRI.NS', 'BIKAJI.NS', 'SUZLON.NS',
                     'CLEAN.NS', 'NYKAA.NS', 'RKFORGE.NS', 'TTML.NS',
                     'AMBER.NS', 'CAPLIPOINT.NS', 'HSCL.NS', 'DEEPAKFERT.NS',
                     'POLICYBZR.NS', 'IEX.NS', 'SAFARI.NS', 'SONACOMS.NS',
                     'ZFCVINDIA.NS', 'GLAXO.NS', 'APLLTD.NS', 'JUBLFOOD.NS',
                     'AJANTPHARM.NS', 'DELHIVERY.NS', 'INDIACEM.NS',
                     'FIVESTAR.NS', 'DIVISLAB.NS', 'STLTECH.NS', 'CASTROLIND.NS',
                     'INOXWIND.NS', 'IDEA.NS', 'PEL.NS', 'JWL.NS', 'GILLETTE.NS',
                     'ASTERDM.NS', 'IDBI.NS', 'JBCHEPHARM.NS', 'CEATLTD.NS',
                     'ANGELONE.NS', 'FDC.NS', 'AFFLE.NS', 'CENTURYPLY.NS',
                     'DEVYANI.NS', 'EQUITASBNK.NS', 'PAGEIND.NS', 'PGHH.NS',
                     'WELCORP.NS', 'MGL.NS', 'ALKYLAMINE.NS', 'MINDACORP.NS',
                     'CHAMBLFERT.NS', 'HDFCBANK.NS', 'BIOCON.NS', 'NBCC.NS',
                     'TORNTPHARM.NS', 'RAMCOCEM.NS', 'AETHER.NS', 'JKPAPER.NS',
                     'LTF.NS', 'HAPPYFORGE.NS', 'DRREDDY.NS', 'UTIAMC.NS',
                     'OIL.NS', 'AUROPHARMA.NS', 'ZEEL.NS', 'PPLPHARMA.NS',
                     'POONAWALLA.NS', 'SUVENPHAR.NS', 'CELLO.NS', 'MUTHOOTFIN.NS',
                     'RAILTEL.NS', 'VARROC.NS', 'PIDILITIND.NS', 'ALKEM.NS',
                     'GLENMARK.NS', 'SUNPHARMA.NS', 'MOTILALOFS.NS',
                     'SCHNEIDER.NS', 'ZYDUSLIFE.NS', 'SHRIRAMFIN.NS',
                     'ARE&M.NS', 'KOTAKBANK.NS', 'BAJAJHLDNG.NS', 'EPL.NS',
                     'ABFRL.NS', 'AAVAS.NS', 'DEEPAKNTR.NS', 'LUPIN.NS',
                     'LLOYDSME.NS', 'UNITDSPR.NS', 'RBA.NS', 'ASIANPAINT.NS',
                     'TORNTPOWER.NS', 'HINDPETRO.NS', 'MFSL.NS', 'ALLCARGO.NS',
                     'CDSL.NS', 'DOMS.NS', 'NESTLEIND.NS', 'CANFINHOME.NS',
                     'GLAND.NS', 'TANLA.NS', 'ASHOKLEY.NS', 'ABBOTINDIA.NS',
                     'MHRIL.NS', 'PRSMJOHNSN.NS', 'POLYMED.NS', 'KPITTECH.NS',
                     'BALRAMCHIN.NS', 'PNBHOUSING.NS', 'COROMANDEL.NS',
                     'ASTRAZEN.NS', 'JINDALSAW.NS', 'METROPOLIS.NS', 'RITES.NS',
                     'GSPL.NS', 'APOLLOHOSP.NS', 'LEMONTREE.NS', 'NATCOPHARM.NS',
                     'VGUARD.NS', 'SUNTV.NS', '3MINDIA.NS', 'RAINBOW.NS',
                     'COLPAL.NS', 'CIPLA.NS', 'DATAPATTNS.NS', 'TRITURBINE.NS',
                     'LALPATHLAB.NS', 'HUDCO.NS', 'BANDHANBNK.NS', 'KPRMILL.NS',
                     'JMFINANCIL.NS', 'MASTEK.NS', 'BRITANNIA.NS', 'RENUKA.NS',
                     'UBL.NS', 'LATENTVIEW.NS', 'CYIENT.NS', 'ECLERX.NS',
                     'APTUS.NS', 'GMMPFAUDLR.NS', 'GUJGASLTD.NS', 'DIXON.NS',
                     'FORTIS.NS', 'KRBL.NS', 'CHEMPLASTS.NS', 'GICRE.NS',
                     'LXCHEM.NS', 'JUSTDIAL.NS', 'BANKINDIA.NS', 'NETWORK18.NS',
                     'AARTIIND.NS', 'SPARC.NS', 'BERGEPAINT.NS', 'RHIM.NS',
                     'SUNDRMFAST.NS', 'NCC.NS', 'GAEL.NS', 'VBL.NS',
                     'TATACOMM.NS', 'UPL.NS', 'CHOLAFIN.NS', 'POLYCAB.NS',
                     'BATAINDIA.NS', 'IGL.NS', 'ADANIPORTS.NS', 'CSBBANK.NS',
                     'UCOBANK.NS', 'INDIANB.NS', 'TATAPOWER.NS', 'PRAJIND.NS',
                     'AIAENG.NS', 'MAZDOCK.NS', 'WHIRL.NS', 'TATAELXSI.NS',
                     'BAYERCROP.NS', 'M&MFIN.NS', 'LTTS.NS', 'GPIL.NS',
                     'HAVELLS.NS', 'TMB.NS', 'CHALET.NS', 'JSL.NS',
                     'JYOTHYLAB.NS', 'DCMSHRIRAM.NS', 'ADANIENSOL.NS',
                     'MSUMI.NS', 'FINCABLES.NS', 'AUBANK.NS', 'SANOFI.NS',
                     'CRISIL.NS', 'SUMICHEM.NS', 'BALAMINES.NS', 'HONAUT.NS',
                     'ATUL.NS', 'SIGNATURE.NS', 'HOMEFIRST.NS', 'BPCL.NS',
                     'ANANDRATHI.NS', 'REDINGTON.NS', 'HFCL.NS', 'INDHOTEL.NS',
                     'MANKIND.NS', 'KIMS.NS', 'TATATECH.NS', 'BORORENEW.NS',
                     'GODFRYPHLP.NS', 'GODREJCP.NS', 'EXIDEIND.NS',
                     'EASEMYTRIP.NS', 'ERIS.NS', 'IRFC.NS', 'JIOFIN.NS',
                     'ITI.NS', 'DALBHARAT.NS', 'ICICIPRULI.NS', 'MAHABANK.NS',
                     'KAJARIACER.NS', 'TATAMOTORS.NS', 'HGS.NS', 'RELAXO.NS',
                     'PERSISTENT.NS', 'FINEORG.NS', 'GSFC.NS', 'EICHERMOT.NS',
                     'TRIVENI.NS', 'GREENPANEL.NS', 'CAMS.NS', 'SAPPHIRE.NS',
                     'SHARDACROP.NS', 'APEX.NS', 'BRIGADE.NS', 'UJJIVANSFB.NS',
                     'PNCINFRA.NS', 'KARURVYSYA.NS', 'MIDHANI.NS', 'AHL.NS',
                     'INDIAMART.NS', 'CROMPTON.NS', 'RVNL.NS', 'SUPREMEIND.NS',
                     'HGS.NS']

In [ ]:
data = yf.download(nifty_500, start='2019-01-01', end='2024-01-01')
# data.to_csv('nifty_test1_data.csv')
# data = pd.read_csv('nifty_test1_data.csv', low_memory=False)
# data.head()

[*********************100%%**********************]  244 of 244 completed
ERROR:yfinance:
2 Failed downloads:
ERROR:yfinance:['WHIRL.NS']: YFTzMissingError('$%ticker%: possibly delisted; No timezone found')
ERROR:yfinance:['UNITDSPR.NS']: YFChartError("%ticker%: Data doesn't exist for startDate = 1546281000, endDate = 1704047400")


In [ ]:
returns = data['Adj Close'].pct_change()

<ipython-input-5-b255c7988ed4>:1: FutureWarning: The default fill_method='pad' in DataFrame.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  returns = data['Adj Close'].pct_change()


In [ ]:
volatility = returns.std() * np.sqrt(252)


In [ ]:
sorted_vol = volatility.sort_values(ascending=False)

print(f"\n NIFTY 500 Stocks based on volatility from highest to lowest: ")
print("High Volatility Stocks: ")
print(sorted_vol.head(50))



 NIFTY 500 Stocks based on volatility from highest to lowest: 
High Volatility Stocks: 
Ticker
INDIAMART.NS     0.889373
IDEA.NS          0.840294
SUZLON.NS        0.687312
TTML.NS          0.638760
ADANIENSOL.NS    0.618202
ZEEL.NS          0.590796
JWL.NS           0.585536
INOXWIND.NS      0.583360
EASEMYTRIP.NS    0.558421
J&KBANK.NS       0.557836
RENUKA.NS        0.557492
GPIL.NS          0.552803
ZOMATO.NS        0.552033
ANGELONE.NS      0.543200
BORORENEW.NS     0.542629
TANLA.NS         0.540972
NETWORK18.NS     0.535190
HFCL.NS          0.534653
ITI.NS           0.533735
AWL.NS           0.531322
JSL.NS           0.530335
NCC.NS           0.525579
PAYTM.NS         0.525514
BALAMINES.NS     0.523385
KRBL.NS          0.523143
POONAWALLA.NS    0.522659
HSCL.NS          0.522635
IDBI.NS          0.514719
TRIVENI.NS       0.510584
APEX.NS          0.509595
RVNL.NS          0.507867
SPARC.NS         0.507296
BANDHANBNK.NS    0.506502
POLICYBZR.NS     0.506397
SCHNEIDER.NS     0.5

In [ ]:
print(sorted_vol.tail(50))

Ticker
PAGEIND.NS       0.315103
AETHER.NS        0.313556
AIAENG.NS        0.312506
3MINDIA.NS       0.311615
NATCOPHARM.NS    0.310344
LUPIN.NS         0.307677
CASTROLIND.NS    0.307563
AJANTPHARM.NS    0.304478
ZYDUSLIFE.NS     0.304195
RAMCOCEM.NS      0.304079
HAVELLS.NS       0.301917
MSUMI.NS         0.301913
ERIS.NS          0.301535
GODREJCP.NS      0.300967
DIVISLAB.NS      0.296632
SIGNATURE.NS     0.294904
ATUL.NS          0.294046
KOTAKBANK.NS     0.291845
VGUARD.NS        0.291234
EXIDEIND.NS      0.290188
COROMANDEL.NS    0.289849
ANANDRATHI.NS    0.286654
TORNTPHARM.NS    0.282189
SUNPHARMA.NS     0.280995
TMB.NS           0.277580
BATAINDIA.NS     0.277322
BERGEPAINT.NS    0.276448
RELAXO.NS        0.275161
CIPLA.NS         0.273078
BAYERCROP.NS     0.272409
UBL.NS           0.271207
ZFCVINDIA.NS     0.268873
HDFCBANK.NS      0.268663
ASIANPAINT.NS    0.263360
ABBOTINDIA.NS    0.261813
DRREDDY.NS       0.257906
ALKEM.NS         0.253266
PIDILITIND.NS    0.252243
BRITA

In [ ]:
# sorted_vol = volatility.sort_values(ascending=False)
# sorted_vol_clean = sorted_vol.dropna()
# print("Low Volatility Stocks: ")
# print(sorted_vol_clean.tail(50))


# sorted_vol = volatility.sort_values(ascending=True)
# sorted_vol_clean = sorted_vol.dropna()
# print("High Volatility Stocks: ")
# print(sorted_vol_clean.head(50))

sorted_vol = volatility.sort_values(ascending=False)
sorted_vol_clean = sorted_vol.dropna()
low_volatility_stocks = sorted_vol_clean.tail(50)[::-1]
print(f"\n NIFTY 500 Stocks based on volatility from lowest to highest:")
print("Low Volatility Stocks: ")
print(low_volatility_stocks)



 NIFTY 500 Stocks based on volatility from lowest to highest:
Low Volatility Stocks: 
Ticker
GILLETTE.NS      0.198795
SANOFI.NS        0.215793
COLPAL.NS        0.221769
PGHH.NS          0.227332
NESTLEIND.NS     0.229931
HAPPYFORGE.NS    0.239206
GLAXO.NS         0.242858
DOMS.NS          0.244477
CELLO.NS         0.246616
BRITANNIA.NS     0.249303
PIDILITIND.NS    0.252243
ALKEM.NS         0.253266
DRREDDY.NS       0.257906
ABBOTINDIA.NS    0.261813
ASIANPAINT.NS    0.263360
HDFCBANK.NS      0.268663
ZFCVINDIA.NS     0.268873
UBL.NS           0.271207
BAYERCROP.NS     0.272409
CIPLA.NS         0.273078
RELAXO.NS        0.275161
BERGEPAINT.NS    0.276448
BATAINDIA.NS     0.277322
TMB.NS           0.277580
SUNPHARMA.NS     0.280995
TORNTPHARM.NS    0.282189
ANANDRATHI.NS    0.286654
COROMANDEL.NS    0.289849
EXIDEIND.NS      0.290188
VGUARD.NS        0.291234
KOTAKBANK.NS     0.291845
ATUL.NS          0.294046
SIGNATURE.NS     0.294904
DIVISLAB.NS      0.296632
GODREJCP.NS      0.300

Market Caps


In [ ]:

market_caps = {}
for ticker in nifty_500:
  try:
    info = yf.Ticker(ticker).info
    market_caps[ticker] = info['marketCap']
  except:
    print(f"Couldn't fetch market cap for {ticker}")

# Print market caps
for ticker, market_cap in market_caps.items():
  print(f"{ticker}: {market_cap}")


ERROR:yfinance:404 Client Error: Not Found for url: https://query2.finance.yahoo.com/v10/finance/quoteSummary/WHIRL.NS?modules=financialData%2CquoteType%2CdefaultKeyStatistics%2CassetProfile%2CsummaryDetail&corsDomain=finance.yahoo.com&formatted=false&symbol=WHIRL.NS&crumb=HJ3pp5vFwB0


Couldn't fetch market cap for WHIRL.NS
ZOMATO.NS: 2301770596352
AWL.NS: 472173740032
J&KBANK.NS: 120447066112
PAYTM.NS: 359079673856
PCBL.NS: 152721539072
JUBLINGREA.NS: 101692407808
NAUKRI.NS: 957844750336
BIKAJI.NS: 212424097792
SUZLON.NS: 1091173548032
CLEAN.NS: 167665991680
NYKAA.NS: 554148757504
RKFORGE.NS: 168517664768
TTML.NS: 174497054720
AMBER.NS: 139798790144
CAPLIPOINT.NS: 120535572480
HSCL.NS: 229435129856
DEEPAKFERT.NS: 119982907392
POLICYBZR.NS: 767669829632
IEX.NS: 173443776512
SAFARI.NS: 117021777920
SONACOMS.NS: 402731040768
ZFCVINDIA.NS: 297679388672
GLAXO.NS: 492320751616
APLLTD.NS: 214145564672
JUBLFOOD.NS: 424088141824
AJANTPHARM.NS: 375167320064
DELHIVERY.NS: 308006223872
INDIACEM.NS: 113096908800
FIVESTAR.NS: 207868084224
DIVISLAB.NS: 1230156726272
STLTECH.NS: 64960790528
CASTROLIND.NS: 249555484672
INOXWIND.NS: 277368274944
IDEA.NS: 1106832785408
PEL.NS: 212556906496
JWL.NS: 233176940544
GILLETTE.NS: 255992594432
ASTERDM.NS: 194778759168
IDBI.NS: 1013951365120
J

In [ ]:

enterprise_values = {}
for ticker in nifty_500:
  try:
    info = yf.Ticker(ticker).info
    market_cap = info.get('marketCap')
    total_debt = info.get('totalDebt')

    cash_and_equivalents = info.get('totalCash')
    if market_cap is not None and total_debt is not None and cash_and_equivalents is not None:
      enterprise_value = market_cap + total_debt - cash_and_equivalents
      enterprise_values[ticker] = enterprise_value
    else:
      print(f"Couldn't fetch all necessary data for {ticker}")
  except:
    print(f"Couldn't fetch data for {ticker}")

# Print enterprise values
for ticker, ev in enterprise_values.items():
  print(f"{ticker}: {ev}")


Couldn't fetch all necessary data for ALLCARGO.NS
Couldn't fetch all necessary data for GICRE.NS


ERROR:yfinance:404 Client Error: Not Found for url: https://query2.finance.yahoo.com/v10/finance/quoteSummary/WHIRL.NS?modules=financialData%2CquoteType%2CdefaultKeyStatistics%2CassetProfile%2CsummaryDetail&corsDomain=finance.yahoo.com&formatted=false&symbol=WHIRL.NS&crumb=HJ3pp5vFwB0


Couldn't fetch all necessary data for WHIRL.NS
Couldn't fetch all necessary data for JIOFIN.NS
ZOMATO.NS: 2265910596608
AWL.NS: 483726440448
J&KBANK.NS: 100593864704
PAYTM.NS: 285043672448
PCBL.NS: 198394937088
JUBLINGREA.NS: 108299007872
NAUKRI.NS: 918806979840
BIKAJI.NS: 212313658880
SUZLON.NS: 1088324047872
CLEAN.NS: 164556211680
NYKAA.NS: 561445047040
RKFORGE.NS: 178291953408
TTML.NS: 374301047680
AMBER.NS: 147211346944
CAPLIPOINT.NS: 112648072480
HSCL.NS: 233277255168
DEEPAKFERT.NS: 155279205888
POLICYBZR.NS: 729043830528
IEX.NS: 164217477224
SAFARI.NS: 114713977856
SONACOMS.NS: 403794330624
ZFCVINDIA.NS: 285165413632
GLAXO.NS: 473939168656
APLLTD.NS: 218075064576
JUBLFOOD.NS: 463608761344
AJANTPHARM.NS: 370940319872
DELHIVERY.NS: 297886093312
INDIACEM.NS: 138426808832
FIVESTAR.NS: 254272430080
DIVISLAB.NS: 1193206727552
STLTECH.NS: 94470790656
CASTROLIND.NS: 237657384192
INOXWIND.NS: 309313775104
IDEA.NS: 3415144775680
PEL.NS: 644158291968
JWL.NS: 234908454656
GILLETTE.NS: 251930

In [ ]:
for ticker, ev in enterprise_values.items():
  print(f"{ticker}, {ev} , {market_cap}")

ZOMATO.NS, 2265910596608 , 33690017792
AWL.NS, 483726440448 , 33690017792
J&KBANK.NS, 100593864704 , 33690017792
PAYTM.NS, 285043672448 , 33690017792
PCBL.NS, 198394937088 , 33690017792
JUBLINGREA.NS, 108299007872 , 33690017792
NAUKRI.NS, 918806979840 , 33690017792
BIKAJI.NS, 212313658880 , 33690017792
SUZLON.NS, 1088324047872 , 33690017792
CLEAN.NS, 164556211680 , 33690017792
NYKAA.NS, 561445047040 , 33690017792
RKFORGE.NS, 178291953408 , 33690017792
TTML.NS, 374301047680 , 33690017792
AMBER.NS, 147211346944 , 33690017792
CAPLIPOINT.NS, 112648072480 , 33690017792
HSCL.NS, 233277255168 , 33690017792
DEEPAKFERT.NS, 155279205888 , 33690017792
POLICYBZR.NS, 729043830528 , 33690017792
IEX.NS, 164217477224 , 33690017792
SAFARI.NS, 114713977856 , 33690017792
SONACOMS.NS, 403794330624 , 33690017792
ZFCVINDIA.NS, 285165413632 , 33690017792
GLAXO.NS, 473939168656 , 33690017792
APLLTD.NS, 218075064576 , 33690017792
JUBLFOOD.NS, 463608761344 , 33690017792
AJANTPHARM.NS, 370940319872 , 33690017792

In [ ]:
# for ticker, ev in enterprise_values.items():
#   print(f"{ticker}, {ev} , {market_cap}, {volatility}")

ZOMATO.NS, 2265910596608 , 33690017792, Ticker
3MINDIA.NS       0.311615
AARTIIND.NS      0.342975
AAVAS.NS         0.386200
ABBOTINDIA.NS    0.261813
ABFRL.NS         0.383697
                   ...   
WHIRL.NS              NaN
ZEEL.NS          0.590796
ZFCVINDIA.NS     0.268873
ZOMATO.NS        0.552033
ZYDUSLIFE.NS     0.304195
Length: 244, dtype: float64
AWL.NS, 483726440448 , 33690017792, Ticker
3MINDIA.NS       0.311615
AARTIIND.NS      0.342975
AAVAS.NS         0.386200
ABBOTINDIA.NS    0.261813
ABFRL.NS         0.383697
                   ...   
WHIRL.NS              NaN
ZEEL.NS          0.590796
ZFCVINDIA.NS     0.268873
ZOMATO.NS        0.552033
ZYDUSLIFE.NS     0.304195
Length: 244, dtype: float64
J&KBANK.NS, 100593864704 , 33690017792, Ticker
3MINDIA.NS       0.311615
AARTIIND.NS      0.342975
AAVAS.NS         0.386200
ABBOTINDIA.NS    0.261813
ABFRL.NS         0.383697
                   ...   
WHIRL.NS              NaN
ZEEL.NS          0.590796
ZFCVINDIA.NS     0.268873
Z

In [ ]:
import pandas as pd
import numpy as np
import yfinance as yf

nifty_500 = ['ZOMATO.NS', 'AWL.NS', 'J&KBANK.NS', 'PAYTM.NS', 'PCBL.NS',
                     'JUBLINGREA.NS', 'NAUKRI.NS', 'BIKAJI.NS', 'SUZLON.NS',
                     'CLEAN.NS', 'NYKAA.NS', 'RKFORGE.NS', 'TTML.NS',
                     'AMBER.NS', 'CAPLIPOINT.NS', 'HSCL.NS', 'DEEPAKFERT.NS',
                     'POLICYBZR.NS', 'IEX.NS', 'SAFARI.NS', 'SONACOMS.NS',
                     'ZFCVINDIA.NS', 'GLAXO.NS', 'APLLTD.NS', 'JUBLFOOD.NS',
                     'AJANTPHARM.NS', 'DELHIVERY.NS', 'INDIACEM.NS',
                     'FIVESTAR.NS', 'DIVISLAB.NS', 'STLTECH.NS', 'CASTROLIND.NS',
                     'INOXWIND.NS', 'IDEA.NS', 'PEL.NS', 'JWL.NS', 'GILLETTE.NS',
                     'ASTERDM.NS', 'IDBI.NS', 'JBCHEPHARM.NS', 'CEATLTD.NS',
                     'ANGELONE.NS', 'FDC.NS', 'AFFLE.NS', 'CENTURYPLY.NS',
                     'DEVYANI.NS', 'EQUITASBNK.NS', 'PAGEIND.NS', 'PGHH.NS',
                     'WELCORP.NS', 'MGL.NS', 'ALKYLAMINE.NS', 'MINDACORP.NS',
                     'CHAMBLFERT.NS', 'HDFCBANK.NS', 'BIOCON.NS', 'NBCC.NS',
                     'TORNTPHARM.NS', 'RAMCOCEM.NS', 'AETHER.NS', 'JKPAPER.NS',
                     'LTF.NS', 'HAPPYFORGE.NS', 'DRREDDY.NS', 'UTIAMC.NS',
                     'OIL.NS', 'AUROPHARMA.NS', 'ZEEL.NS', 'PPLPHARMA.NS',
                     'POONAWALLA.NS', 'SUVENPHAR.NS', 'CELLO.NS', 'MUTHOOTFIN.NS',
                     'RAILTEL.NS', 'VARROC.NS', 'PIDILITIND.NS', 'ALKEM.NS',
                     'GLENMARK.NS', 'SUNPHARMA.NS', 'MOTILALOFS.NS',
                     'SCHNEIDER.NS', 'ZYDUSLIFE.NS', 'SHRIRAMFIN.NS',
                     'ARE&M.NS', 'KOTAKBANK.NS', 'BAJAJHLDNG.NS', 'EPL.NS',
                     'ABFRL.NS', 'AAVAS.NS', 'DEEPAKNTR.NS', 'LUPIN.NS',
                     'LLOYDSME.NS', 'UNITDSPR.NS', 'RBA.NS', 'ASIANPAINT.NS',
                     'TORNTPOWER.NS', 'HINDPETRO.NS', 'MFSL.NS', 'ALLCARGO.NS',
                     'CDSL.NS', 'DOMS.NS', 'NESTLEIND.NS', 'CANFINHOME.NS',
                     'GLAND.NS', 'TANLA.NS', 'ASHOKLEY.NS', 'ABBOTINDIA.NS',
                     'MHRIL.NS', 'PRSMJOHNSN.NS', 'POLYMED.NS', 'KPITTECH.NS',
                     'BALRAMCHIN.NS', 'PNBHOUSING.NS', 'COROMANDEL.NS',
                     'ASTRAZEN.NS', 'JINDALSAW.NS', 'METROPOLIS.NS', 'RITES.NS',
                     'GSPL.NS', 'APOLLOHOSP.NS', 'LEMONTREE.NS', 'NATCOPHARM.NS',
                     'VGUARD.NS', 'SUNTV.NS', '3MINDIA.NS', 'RAINBOW.NS',
                     'COLPAL.NS', 'CIPLA.NS', 'DATAPATTNS.NS', 'TRITURBINE.NS',
                     'LALPATHLAB.NS', 'HUDCO.NS', 'BANDHANBNK.NS', 'KPRMILL.NS',
                     'JMFINANCIL.NS', 'MASTEK.NS', 'BRITANNIA.NS', 'RENUKA.NS',
                     'UBL.NS', 'LATENTVIEW.NS', 'CYIENT.NS', 'ECLERX.NS',
                     'APTUS.NS', 'GMMPFAUDLR.NS', 'GUJGASLTD.NS', 'DIXON.NS',
                     'FORTIS.NS', 'KRBL.NS', 'CHEMPLASTS.NS', 'GICRE.NS',
                     'LXCHEM.NS', 'JUSTDIAL.NS', 'BANKINDIA.NS', 'NETWORK18.NS',
                     'AARTIIND.NS', 'SPARC.NS', 'BERGEPAINT.NS', 'RHIM.NS',
                     'SUNDRMFAST.NS', 'NCC.NS', 'GAEL.NS', 'VBL.NS',
                     'TATACOMM.NS', 'UPL.NS', 'CHOLAFIN.NS', 'POLYCAB.NS',
                     'BATAINDIA.NS', 'IGL.NS', 'ADANIPORTS.NS', 'CSBBANK.NS',
                     'UCOBANK.NS', 'INDIANB.NS', 'TATAPOWER.NS', 'PRAJIND.NS',
                     'AIAENG.NS', 'MAZDOCK.NS', 'WHIRL.NS', 'TATAELXSI.NS',
                     'BAYERCROP.NS', 'M&MFIN.NS', 'LTTS.NS', 'GPIL.NS',
                     'HAVELLS.NS', 'TMB.NS', 'CHALET.NS', 'JSL.NS',
                     'JYOTHYLAB.NS', 'DCMSHRIRAM.NS', 'ADANIENSOL.NS',
                     'MSUMI.NS', 'FINCABLES.NS', 'AUBANK.NS', 'SANOFI.NS',
                     'CRISIL.NS', 'SUMICHEM.NS', 'BALAMINES.NS', 'HONAUT.NS',
                     'ATUL.NS', 'SIGNATURE.NS', 'HOMEFIRST.NS', 'BPCL.NS',
                     'ANANDRATHI.NS', 'REDINGTON.NS', 'HFCL.NS', 'INDHOTEL.NS',
                     'MANKIND.NS', 'KIMS.NS', 'TATATECH.NS', 'BORORENEW.NS',
                     'GODFRYPHLP.NS', 'GODREJCP.NS', 'EXIDEIND.NS',
                     'EASEMYTRIP.NS', 'ERIS.NS', 'IRFC.NS', 'JIOFIN.NS',
                     'ITI.NS', 'DALBHARAT.NS', 'ICICIPRULI.NS', 'MAHABANK.NS',
                     'KAJARIACER.NS', 'TATAMOTORS.NS', 'HGS.NS', 'RELAXO.NS',
                     'PERSISTENT.NS', 'FINEORG.NS', 'GSFC.NS', 'EICHERMOT.NS',
                     'TRIVENI.NS', 'GREENPANEL.NS', 'CAMS.NS', 'SAPPHIRE.NS',
                     'SHARDACROP.NS', 'APEX.NS', 'BRIGADE.NS', 'UJJIVANSFB.NS',
                     'PNCINFRA.NS', 'KARURVYSYA.NS', 'MIDHANI.NS', 'AHL.NS',
                     'INDIAMART.NS', 'CROMPTON.NS', 'RVNL.NS', 'SUPREMEIND.NS',
                     'HGS.NS']

results = []

# Loop through each stock to calculate volatility, variance, fetch enterprise value, operating margin, and earnings per share
for ticker in nifty_500:
    try:
        # Fetch stock info
        stock = yf.Ticker(ticker)
        info = stock.info

        # Debugging: Print the fetched info for each stock
        print(f"Info for {ticker}: {info}")

        # Calculate volatility from daily returns
        data = stock.history(period="max")  # Get max available data
        returns = data['Close'].pct_change().dropna()  # Calculate daily returns

        # Calculate volatility (annualized)
        volatility = returns.std() * np.sqrt(252)

        # Calculate variance (daily)
        variance = returns.var()

        # Get enterprise value components
        market_cap = info.get('marketCap')
        total_debt = info.get('totalDebt')
        cash_and_equivalents = info.get('totalCash')

        # Calculate enterprise value
        if market_cap is not None and total_debt is not None and cash_and_equivalents is not None:
            enterprise_value = market_cap + total_debt - cash_and_equivalents
        else:
            enterprise_value = None

        # Fetch operating margin
        operating_margin = info.get('operatingMargins')

        # Fetch earnings per share (EPS)
        earnings_per_share = info.get('epsTrailingTwelveMonths')

        # Calculate Enterprise Value / Operating Margin ratio
        if enterprise_value is not None and operating_margin not in [None, 0]:
            ev_to_operating_margin = enterprise_value / operating_margin
        else:
            ev_to_operating_margin = None

        # Append the result to the DataFrame
        results.append({
            'Ticker': ticker,
            'Enterprise Value': enterprise_value,
            'Market Cap': market_cap,
            'Volatility': volatility,
            'Variance': variance,
            'Operating Margin': operating_margin,
            'Earnings Per Share': earnings_per_share,
            'EV/Operating Margin': ev_to_operating_margin
        })

    except Exception as e:
        print(f"Couldn't fetch data for {ticker}: {e}")

# Convert results to DataFrame
results_df = pd.DataFrame(results)

# Print or save the results
print(results_df.head())

# Save the results to CSV if needed
results_df.to_csv('Nifty_500_with_yfinfo.csv', index=False)


Info for ZOMATO.NS: {'address1': 'Pioneer Square Building', 'address2': 'Tower 1 - Ground to 6th Floor Near Golf Course Extension, Sector 62', 'city': 'Gurugram', 'zip': '122008', 'country': 'India', 'phone': '91 12 4426 8565', 'website': 'https://www.zomato.com', 'industry': 'Internet Retail', 'industryKey': 'internet-retail', 'industryDisp': 'Internet Retail', 'sector': 'Consumer Cyclical', 'sectorKey': 'consumer-cyclical', 'sectorDisp': 'Consumer Cyclical', 'longBusinessSummary': 'Zomato Limited primarily operates as an online food delivery company in India and internationally. It operates through India Food Ordering and Delivery; Hyperpure Supplies; Quick Commerce; Going Out, and All Other segments. The company operates operates a B2C technology platform under Zomato brand name, that helps to search and discover restaurants, order food delivery, read and write customer generated reviews, view and upload photos, and book a table and make payments while dining-out at restaurants, as 

ERROR:yfinance:404 Client Error: Not Found for url: https://query2.finance.yahoo.com/v10/finance/quoteSummary/WHIRL.NS?modules=financialData%2CquoteType%2CdefaultKeyStatistics%2CassetProfile%2CsummaryDetail&corsDomain=finance.yahoo.com&formatted=false&symbol=WHIRL.NS&crumb=3xxXP..LW9a


Info for WHIRL.NS: {'trailingPegRatio': None}


ERROR:yfinance:$WHIRL.NS: possibly delisted; No timezone found


Couldn't fetch data for WHIRL.NS: attempt to get argmax of an empty sequence
Info for TATAELXSI.NS: {'address1': 'ITPB Road', 'address2': 'Whitefield', 'city': 'Bengaluru', 'zip': '560048', 'country': 'India', 'phone': '91 80 2297 9123', 'fax': '91 80 2841 1474', 'website': 'https://www.tataelxsi.com', 'industry': 'Software - Application', 'industryKey': 'software-application', 'industryDisp': 'Software - Application', 'sector': 'Technology', 'sectorKey': 'technology', 'sectorDisp': 'Technology', 'longBusinessSummary': 'Tata Elxsi Limited engages in the provision of product design and engineering, and systems integration and support services in India, the United States, Europe, and internationally. The company operates through two segments, Systems Integration and Support, and Software Development and Services. It offers AIVA, an AI-based video analytics platform for content curation and object action event meta-tagging; AI DMS, a driver intelligence service solution; ICX, a SaaS-based

'address1': 'Pioneer Square Building', 'address2': 'Tower 1 - Ground to 6th Floor Near Golf Course Extension, Sector 62', 'city': 'Gurugram', 'zip': '122008', 'country': 'India', 'phone': '91 12 4426 8565', 'website': 'https://www.zomato.com', 'industry': 'Internet Retail', 'industryKey': 'internet-retail', 'industryDisp': 'Internet Retail', 'sector': 'Consumer Cyclical', 'sectorKey': 'consumer-cyclical', 'sectorDisp': 'Consumer Cyclical', 'longBusinessSummary': 'Zomato Limited primarily operates as an online food delivery company in India and internationally. It operates through India Food Ordering and Delivery; Hyperpure Supplies; Quick Commerce; Going Out, and All Other segments. The company operates operates a B2C technology platform under Zomato brand name, that helps to search and discover restaurants, order food delivery, read and write customer generated reviews, view and upload photos, and book a table and make payments while dining-out at restaurants, as well as discovery and ticketing services primarily for events, such as food carnivals, music concerts, comedy shows, and others. Its technology platform connects customers, restaurant partners, and delivery partners. The company also operates Hyperpure, a procurement solution that supplies ingredients and kitchen products to restaurant partners; and Blinkit, a quick commerce marketplace delivering everyday products to customers within minutes. In addition, it engages in the provision of event organizing, and payment aggregator and gateway services; and engages in the trading, financing, and investment activities. The company was incorporated in 2010 and is headquartered in Gurugram, India.', 'fullTimeEmployees': 3988, 'companyOfficers': [{'maxAge': 1, 'name': 'Mr. Deepinder  Goyal', 'age': 40, 'title': 'Co-Founder, MD, CEO & Director', 'yearBorn': 1983, 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Mr. Akshant  Goyal', 'title': 'Chief Financial Officer', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Ms. Sandhya  Sethia', 'title': 'Company Secretary & Compliance Officer', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Ms. Akriti  Chopra', 'title': 'Co-founder & Chief People Officer', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Ms. Daminee  Sawhney', 'title': 'Vice President of Operations & People Team', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Ms. Damini  Bhalla', 'title': 'General Counsel', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Deepak  Ahluwalia', 'title': 'Head of Governance, Risk, Compliance & Internal audit', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Kunal  Swarup', 'title': 'Head of Corporate Development', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Ms. Anjalli Ravi Kumar', 'title': 'Chief Sustainability Officer', 'exercisedValue': 0, 'unexercisedValue': 0}, {'maxAge': 1, 'name': 'Rinshul  Chandra', 'title': 'COO of Food Ordering & Delivery Vertical', 'exercisedValue': 0, 'unexercisedValue': 0}], 'maxAge': 86400, 'priceHint': 2, 'previousClose': 264.43, 'open': 278.0, 'dayLow': 266.5, 'dayHigh': 280.9, 'regularMarketPreviousClose': 264.43, 'regularMarketOpen': 278.0, 'regularMarketDayLow': 266.5, 'regularMarketDayHigh': 280.9, 'beta': 0.257, 'trailingPE': 387.0, 'forwardPE': 97.81319, 'volume': 51974580, 'regularMarketVolume': 51974580, 'averageVolume': 54805376, 'averageVolume10days': 75752660, 'averageDailyVolume10Day': 75752660, 'bid': 267.03, 'ask': 267.03, 'marketCap': 2324402798592, 'fiftyTwoWeekLow': 88.3, 'fiftyTwoWeekHigh': 280.9, 'priceToSalesTrailing12Months': 16.717512, 'fiftyDayAverage': 219.759, 'twoHundredDayAverage': 172.50075, 'currency': 'INR', 'enterpriseValue': 2270993842176, 'profitMargins': 0.0433, 'floatShares': 6434836897, 'sharesOutstanding': 8702350336, 'heldPercentInsiders': 0.27643, 'heldPercentInstitutions': 0.40596, 'impliedSharesOutstanding': 8816979968, 'bookValue': 23.329, 'priceToBook': 11.446268, 'lastFiscalYearEnd': 1711843200, 'nextFiscalYearEnd': 1743379200, 'mostRecentQuarter': 1719705600, 'earningsQuarterlyGrowth': 125.5, 'netIncomeToCommon': 6019999744, 'trailingEps': 0.69, 'forwardEps': 2.73, 'pegRatio': 12.04, 'enterpriseToRevenue': 16.333, 'enterpriseToEbitda': 748.145, '52WeekChange': 1.9462953, 'SandP52WeekChange': 0.2623955, 'exchange': 'NSI', 'quoteType': 'EQUITY', 'symbol': 'ZOMATO.NS', 'underlyingSymbol': 'ZOMATO.NS', 'shortName': 'ZOMATO LIMITED', 'longName': 'Zomato Limited', 'firstTradeDateEpochUtc': 1627011900, 'timeZoneFullName': 'Asia/Kolkata', 'timeZoneShortName': 'IST', 'uuid': '99bc8fa2-0d3e-3451-bb32-9175776c2028', 'messageBoardId': 'finmb_112318630', 'gmtOffSetMilliseconds': 19800000, 'currentPrice': 267.03, 'targetHighPrice': 300.0, 'targetLowPrice': 100.0, 'targetMeanPrice': 226.25, 'targetMedianPrice': 230.0, 'recommendationMean': 1.7, 'recommendationKey': 'buy', 'numberOfAnalystOpinions': 24, 'totalCash': 43349999616, 'totalCashPerShare': 4.969, 'ebitda': 3035500032, 'totalDebt': 7489999872, 'totalRevenue': 139039997952, 'debtToEquity': 3.67, 'revenuePerShare': 15.937, 'revenueGrowth': 0.741, 'grossMargins': 0.62104, 'ebitdaMargins': 0.02183, 'operatingMargins': 0.00666, 'financialCurrency': 'INR', 'trailingPegRatio': None